src/toolbox/shamir_secrets.rs

```diff
@@ -25,7 +94,156 @@ impl SecretShare {
         })
     }
 
+    // given enough shares, output the secret
     pub fn reconstruct(shares: Vec<Scalar>) -> Result<Scalar, String> {
-        return Ok(shares[0]); //TODO Shamir Secret Reconstruction
+        // we need to know t, the threshold for shares
+        // do we assume len(shares)?
+        // I think this is safe, since we're the ones setting up the shares in the first place
+
+        // From Handbook of Applied Cryptography, 12.71:
+        //     S = sum(i=1..t){c_i*y_i}, where c_i = TT(j=1..t, j!=i){x_j / (x_j - x_i)}
+        let mut sum: Scalar = Scalar::zero();
+
+        for i in 0..shares.len() {
+            let y_i = shares[i];
+            // println!("Using point ({}, {:?})", i+1, y_i);
+
+            let mut c_i = 1f64;
+
+            for j in 0..shares.len() {
+                if i == j { continue; }
+                c_i *= ((j+1) as f64) / ((j+1) as f64 - (i+1) as f64);
+            }
+
+            // println!("c_{} = {}", i+1, c_i);
+
+            // The From trait on Scalar only supports unsigned numerical types, so we have to kinda hack this
+            // println!("c_{} as u64 = {}", i+1, (-1.0 * c_i) as u64);
+            let c_i = if c_i < 0.0 { Scalar::from((-1.0 * c_i) as u64).neg() } else { Scalar::from(c_i as u64) };
+
+            
+            let delta = c_i * y_i;
+            // println!("Delta is {:?}", delta);
+            sum += delta;
+            // println!("Sum is now {:?}", sum);
+        }
+
+        return Ok(sum);
+    }
+
+    // given enough shares, output the secret
+    pub fn reconstruct_at_index(idx: usize, shares: Vec<Scalar>) -> Result<Scalar, String> {
+        // From Handbook of Applied Cryptography, 12.71:
+        //     S = sum(i=1..t){c_i*y_i}, where c_i = TT(j=1..t, j!=i){x_j / (x_j - x_i)}
+        let mut sum: Scalar = Scalar::zero();
+        for i in 0..shares.len() {
```

Comment:
Loop over shares themselves instead of with an index:
```
for (i, y_i) in shares.iter().enumerate() {

}
```

Reply:
Ah, I didn't know you could use an iterator and get the index back.  Thanks!